File: app/src/main/java/com/yvrun/officeprocess/mvp/net/SimpleObserver.java
package com.yvrun.officeprocess.mvp.net;

import android.util.Log;

import com.yvrun.officeprocess.mvp.view.IBaseView;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public abstract class SimpleObserver<T> implements Observer<T> {

    @Override
    public void onSubscribe(Disposable d) {

        Log.d("--TAG--","onSubscribe");
    }

    @Override
    public void onNext(T t) {

        Log.d("--TAG--","onNext");

        success(t);
    }

    @Override
    public void onError(Throwable e) {

        Log.d("--TAG--","onError");
        error(e);
    }

    @Override
    public void onComplete() {

        Log.d("--TAG--","onComplete");
    }

    protected abstract void success(T t);
    protected abstract void error(Throwable e);

}

File: kotlinappother/src/main/java/com/chengzhen/kotlinappother/mvp/net/SimpleObserver.kt
package com.chengzhen.kotlinappother.mvp.net

import android.util.Log

import io.reactivex.Observer
import io.reactivex.disposables.Disposable

abstract class SimpleObserver<T> : Observer<T> {

    override fun onSubscribe(d: Disposable) {

        Log.d("--TAG--", "onSubscribe")
    }

    override fun onNext(t: T) {

        Log.d("--TAG--", "onNext")

        success(t)
    }

    override fun onError(e: Throwable) {

        Log.d("--TAG--", "onError")
        error(e)
    }

    override fun onComplete() {

        Log.d("--TAG--", "onComplete")
    }

    protected abstract fun success(t: T)
    protected abstract fun error(e: Throwable)

}
